--- s3curl.pl	2010-11-12 13:04:14.000000000 -0800
+++ s3curl-euca.pl	2011-07-13 11:06:13.143976835 -0700
@@ -51,6 +51,7 @@
 my $debug = 0;
 my $copySourceObject;
 my $postBody;
+my $url = "";
 
 my $DOTFILENAME=".s3curl";
 my $EXECFILE=$FindBin::Bin;
@@ -83,6 +84,7 @@
     'contentMd5=s' => \$contentMD5,
     'put=s' => \$fileToPut,
     'copySrc=s' => \$copySourceObject,
+    'url=s' => \$url,
     'post:s' => \$postBody,
     'delete' => \$doDelete,
     'createBucket:s' => \$createBucket,
@@ -91,12 +93,16 @@
     'debug' => \$debug
 );
 
+# if we specified an S3 endpoint, then override the defaults
+@endpoints = ($url) if ($url);
+
 my $usage = <<USAGE;
 Usage $0 --id friendly-name (or AWSAccessKeyId) [options] -- [curl-options] [URL]
  options:
   --key SecretAccessKey       id/key are AWSAcessKeyId and Secret (unsafe)
   --contentType text/plain    set content-type header
   --acl public-read           use a 'canned' ACL (x-amz-acl header)
+  --url s3_url                use 's3_url' as endpoint instead of defaults
   --contentMd5 content_md5    add x-amz-content-md5 header
   --put <filename>            PUT request (from the provided local file)
   --post [<filename>]         POST request (optional local file)
@@ -121,6 +127,7 @@
         "Do you need to set it up in $DOTFILE?"
         unless defined $keyinfo;
 
+    @endpoints = ( $keyinfo->{url} ) if ($keyinfo->{url});
     $keyId = $keyinfo->{id};
     $secretKey = $keyinfo->{key};
 }
@@ -289,12 +296,14 @@
 \%awsSecretAccessKeys = (
     # personal account
     personal => {
+        url => '10.10.10.10',
         id => '1ME55KNV6SBTR7EXG0R2',
         key => 'zyMrlZUKeG9UcYpwzlPko/+Ciu0K2co0duRM3fhi',
     },
 
     # corporate account
     company => {
+        url => '192.168.7.9',
         id => '1ATXQ3HHA59CYF1CVS02',
         key => 'WQY4SrSS95pJUT95V6zWea01gBKBCL6PI0cdxeH8',
     },
